# **PERKENALAN**

## 1. Latar Belakang
Project 1 mengangkat kasus regresi dengan dataset taxi online, pada project ini penulis ingin melihat hubungan dari nilai harga layanan taxi online dengan variabel-variabel yang disediakan (seperti provider taxi online, temperature hari pemesanan, tanggal pemesanan, dll) 

-----------------------------------------------------------------------

## 2. Dataset
### 2.1 Penjelasan dan Sumber
Dataset yang digunakan bersumber dari situs kaggle, Data tersebut merupakan data pemesanan taxi online dari dua provider taxi online terbesar, yaitu uber dan lyft, dataset terdiri dari 693071 data yang terbagi ke 57 kolom.

### 2.2 Penjelasan kelas dan kolom
Kolom-kolom tersebut nantinya akan dijadikan variabel independen dan variabel dependen, dimana nantinya variabel dependen (yaitu kolom harga) akan bergantung nilainya terhadap variabel independen (yaitu kolom numerik selain kolom harga)

-----------------------------------------------------------------------

## 3. Objective
Tujuan dari project ini adalah melihat variabel atau kolom apa yang mempunyai keterikatan kuat terhadap harga layanan dari taxi online


===================================================================================================================

# **IMPORT LIBRARY**

In [1]:
import pandas as pd
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
import sklearn.metrics as skmet

from sklearn.preprocessing import LabelEncoder, StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split

===================================================================================================================

# **DATA LOADING**

In [2]:
df = pd.read_csv('./datasets/rideshare_kaggle.csv')
pd.set_option('display.max_columns', None)
df.head(5)

id     timestamp  hour  day  month  \
0  424553bb-7174-41ea-aeb4-fe06d4f4b9d7  1.544953e+09     9   16     12   
1  4bd23055-6827-41c6-b23b-3c491f24e74d  1.543284e+09     2   27     11   
2  981a3613-77af-4620-a42a-0c0866077d1e  1.543367e+09     1   28     11   
3  c2d88af2-d278-4bfd-a8d0-29ca77cc5512  1.543554e+09     4   30     11   
4  e0126e1f-8ca9-4f2e-82b3-50505a09db9a  1.543463e+09     3   29     11   

              datetime          timezone            source    destination  \
0  2018-12-16 09:30:07  America/New_York  Haymarket Square  North Station   
1  2018-11-27 02:00:23  America/New_York  Haymarket Square  North Station   
2  2018-11-28 01:00:22  America/New_York  Haymarket Square  North Station   
3  2018-11-30 04:53:02  America/New_York  Haymarket Square  North Station   
4  2018-11-29 03:49:20  America/New_York  Haymarket Square  North Station   

  cab_type    product_id          name  price  distance  surge_multiplier  \
0     Lyft     lyft_line        Shared    5.0      0.44               1.0   
1     Lyft  lyft_premier           Lux   11.0      0.44               1.0   
2     Lyft          lyft          Lyft    7.0      0.44               1.0   
3     Lyft   lyft_luxsuv  Lux Black XL   26.0      0.44               1.0   
4     Lyft     lyft_plus       Lyft XL    9.0      0.44               1.0   

   latitude  longitude  temperature  apparentTemperature    short_summary  \
0   42.2148    -71.033        42.34                37.12   Mostly Cloudy    
1   42.2148    -71.033        43.58                37.35            Rain    
2   42.2148    -71.033        38.33                32.93           Clear    
3   42.2148    -71.033        34.38                29.63           Clear    
4   42.2148    -71.033        37.44                30.88   Partly Cloudy    

                                        long_summary  precipIntensity  \
0                          Rain throughout the day.            0.0000   
1   Rain until morning, starting again in the eve...           0.1299   
2                        Light rain in the morning.            0.0000   
3                 Partly cloudy throughout the day.            0.0000   
4                 Mostly cloudy throughout the day.            0.0000   

   precipProbability  humidity  windSpeed  windGust  windGustTime  visibility  \
0                0.0      0.68       8.66      9.17    1545015600      10.000   
1                1.0      0.94      11.98     11.98    1543291200       4.786   
2                0.0      0.75       7.33      7.33    1543334400      10.000   
3                0.0      0.73       5.28      5.28    1543514400      10.000   
4                0.0      0.70       9.14      9.14    1543446000      10.000   

   temperatureHigh  temperatureHighTime  temperatureLow  temperatureLowTime  \
0            43.68           1544968800           34.19          1545048000   
1            47.30           1543251600           42.10          1543298400   
2            47.55           1543320000           33.10          1543402800   
3            45.03           1543510800           28.90          1543579200   
4            42.18           1543420800           36.71          1543478400   

   apparentTemperatureHigh  apparentTemperatureHighTime  \
0                    37.95                   1544968800   
1                    43.92                   1543251600   
2                    44.12                   1543320000   
3                    38.53                   1543510800   
4                    35.75                   1543420800   

   apparentTemperatureLow  apparentTemperatureLowTime                   icon  \
0                   27.39                  1545044400   partly-cloudy-night    
1                   36.20                  1543291200                  rain    
2                   29.11                  1543392000           clear-night    
3                   26.20                  1543575600           clear-night    
4                   30.29                

#### Jumlah Data

In [3]:
print("Jumlah Data : ", len(df))

Jumlah Data :  693071


#### Deskripsi Dataset

In [4]:
df.describe()

timestamp           hour            day          month  \
count  6.930710e+05  693071.000000  693071.000000  693071.000000   
mean   1.544046e+09      11.619137      17.794365      11.586684   
std    6.891925e+05       6.948114       9.982286       0.492429   
min    1.543204e+09       0.000000       1.000000      11.000000   
25%    1.543444e+09       6.000000      13.000000      11.000000   
50%    1.543737e+09      12.000000      17.000000      12.000000   
75%    1.544828e+09      18.000000      28.000000      12.000000   
max    1.545161e+09      23.000000      30.000000      12.000000   

               price       distance  surge_multiplier       latitude  \
count  637976.000000  693071.000000     693071.000000  693071.000000   
mean       16.545125       2.189430          1.013870      42.338172   
std         9.324359       1.138937          0.091641       0.047840   
min         2.500000       0.020000          1.000000      42.214800   
25%         9.000000       1.280000          1.000000      42.350300   
50%        13.500000       2.160000          1.000000      42.351900   
75%        22.500000       2.920000          1.000000      42.364700   
max        97.500000       7.860000          3.000000      42.366100   

           longitude    temperature  apparentTemperature  precipIntensity  \
count  693071.000000  693071.000000        693071.000000    693071.000000   
mean      -71.066151      39.584388            35.881222         0.008922   
std         0.020302       6.726084             7.918707         0.026901   
min       -71.105400      18.910000            12.130000         0.000000   
25%       -71.081000      36.450000            31.910000         0.000000   
50%       -71.063100      40.490000            35.900000         0.000000   
75%       -71.054200      43.580000            40.080000         0.000000   
max       -71.033000      57.220000            57.220000         0.144700   

       precipProbability       humidity      windSpeed       windGust  \
count      693071.000000  693071.000000  693071.000000  693071.000000   
mean            0.146076       0.741119       6.186253       8.469860   
std             0.328938       0.138595       3.148223       5.289179   
min             0.000000       0.380000       0.450000       0.800000   
25%             0.000000       0.640000       3.410000       4.060000   
50%             0.000000       0.710000       5.910000       7.550000   
75%             0.000000       0.880000       8.410000      11.740000   
max             1.000000       0.960000      15.000000      27.250000   

       windGustTime     visibility  temperatureHigh  temperatureHighTime  \
count  6.930710e+05  693071.000000    693071.000000         6.930710e+05   
mean   1.544049e+09       8.467970        45.040982         1.544050e+09   
std    6.928244e+05       2.602888         5.996541         6.937921e+05   
min    1.543151e+09       0.717000        32.680000         1.543154e+09   
25%    1.543432e+09       8.432000        42.570000         1.543439e+09   
50%    1.543756e+09       9.880000        44.680000         1.543788e+09   
75%    1.544846e+09       9.996000        46.910000         1.544814e+09   
max    1.545127e+09      10.000000        57.870000         1.545160e+09   

       temperatureLow  temperatureLowTime  apparentTemperatureHigh  \
count   693071.000000        6.930710e+05            693071.000000   
mean        34.150736        1.544102e+09                41.614054   
std          6.383163        6.922923e+05                 7.666138   
min         17.850000        1.543234e+09                22.620000   
25%         30.170000        1.543489e+09                36.570000   
50%         34.180000        1.543817e+09                40.950000   
75%         38.730000        1.544836e+09                44.120000   
max         46.600000        1.545221e+09                57.200000   

       apparentTemperatureHighTime  apparentTemperatureLow  \
count             

#### Kolom dataset

In [5]:
print(df.columns)

Index(['id', 'timestamp', 'hour', 'day', 'month', 'datetime', 'timezone',
       'source', 'destination', 'cab_type', 'product_id', 'name', 'price',
       'distance', 'surge_multiplier', 'latitude', 'longitude', 'temperature',
       'apparentTemperature', 'short_summary', 'long_summary',
       'precipIntensity', 'precipProbability', 'humidity', 'windSpeed',
       'windGust', 'windGustTime', 'visibility', 'temperatureHigh',
       'temperatureHighTime', 'temperatureLow', 'temperatureLowTime',
       'apparentTemperatureHigh', 'apparentTemperatureHighTime',
       'apparentTemperatureLow', 'apparentTemperatureLowTime', 'icon',
       'dewPoint', 'pressure', 'windBearing', 'cloudCover', 'uvIndex',
       'visibility.1', 'ozone', 'sunriseTime', 'sunsetTime', 'moonPhase',
       'precipIntensityMax', 'uvIndexTime', 'temperatureMin',
       'temperatureMinTime', 'temperatureMax', 'temperatureMaxTime',
       'apparentTemperatureMin', 'apparentTemperatureMinTime',
       'apparentTemperat

#### Tipe Data Tiap Kolom

In [6]:
display(df.dtypes)

id                              object
timestamp                      float64
hour                             int64
day                              int64
month                            int64
datetime                        object
timezone                        object
source                          object
destination                     object
cab_type                        object
product_id                      object
name                            object
price                          float64
distance                       float64
surge_multiplier               float64
latitude                       float64
longitude                      float64
temperature                    float64
apparentTemperature            float64
short_summary                   object
long_summary                    object
precipIntensity                float64
precipProbability              float64
humidity                       float64
windSpeed                      float64
windGust                 

===================================================================================================================

# **DATA CLEANING**

#### Handling Missing Value

In [ ]:
df.isnull().sum()

Penjelasan : Kolom price yang nantinya akan dijadikan variabel dependen mempunyai missing value sebanyak 55095 data

Solusi : Data missing value tersebut akan didrop, solusi ini dipilih karena data yang missing tersebut jumlahnya tidak terlalu banyak dibandingkan jumlah total data yaitu 693071 banding 55095 data

---

#### Penghapusan data missing value

In [ ]:
df.dropna(axis=0, inplace=True)
df['price'].isnull().sum()

In [ ]:
print("Jumlah data sekarang : ", len(df))

Seperti yang bisa dilihat jumlah nilai missing value pada kolom price telah habis

Dan jumlah data setelah dilakukan handling missing value berubah dari  693071 data menjadi 637976 data, tidak terlalu banyak data yang terbuang

---

#### Pengecekan kolom dengan nilai kategorikal dan numerik

In [ ]:
display(df.dtypes)

In [ ]:
df['cab_type'].value_counts()

Kolom yang ,mempunyai tipe data object atau string akan didrop, kecuali kolom "cab_type", karena hanya memiliki 2 nilai yang nantinya bisa diubah kedalam bentuk numerik

---

#### Pengubahan nilai kolom "cab_type" menjadi nilai numerik

In [ ]:
le = LabelEncoder()

In [ ]:
df['cab_type'] = le.fit_transform(df.cab_type)

In [ ]:
df['cab_type']

---

#### Penghapusan Kolom yang tidak dipakai

In [ ]:
display(df.dtypes)

In [ ]:
df = df.drop(['id','timestamp' ,'datetime', 'timezone', 'source', 'destination', 'product_id', 'name', 'short_summary', 'long_summary', 'icon'], axis=1)

In [ ]:
df

In [ ]:
df.surge_multiplier.unique()

---

# Data Preprocessing

Karena kolomnya sangat banyak, jadinya perlu diambil kolom mana yang paling berpengaruh terhadap price. Untuk itu, dilakukanlah pencarian korelasi.

In [ ]:
dfcorr = df.corr()
dfcorr

Berikutnya melihat korelasi dengan heatmap

In [ ]:
plt.figure(figsize=(20, 20))
sns.heatmap(dfcorr, cmap='ocean')

Tahap berikutnya menyeleksi kolom mana yang ingin dipakai dengan acuan nilai korelasinya di atas 0,005 dan di bawah -0,005 terhadap price. Alasannya adalah semakin angka tersebut mendekati 0 maka pengaruhnya makin kecil dan price-lah yang ingin dijadikan sebagai hasil forecasting untuk regresi linear

In [ ]:
dfcorr2 = dfcorr[(dfcorr['price'] >= 0.001) | (dfcorr['price'] <= -0.001)]
dfcorr2['price']

In [ ]:
dfcorr3 = dfcorr[(dfcorr['surge_multiplier'] >= 0.001) | (dfcorr['surge_multiplier'] <= -0.001)]
dfcorr3['surge_multiplier']

Setelah melakukan penyaringan dengan patokan korelasi, maka kita akan mengambil 8 kolom yaitu distance,surge_multiplier, price, cab_type, windgust, windbearing, visibility, dan moonphase. Sebelumnya, mentor menyarankan untuk mengambil hour dan day sebagai patokan juga. Akan tetapi, korelasinya sedikit sehingga tidak mengambil hour dan day.

In [ ]:
true_df = df[['distance', 'surge_multiplier', 'price', 'cab_type', 'visibility', 'windBearing', 'moonPhase', 'windGust']]
true_df.head(3)

In [ ]:
plt.figure(figsize=(6, 6))
plt.title('Korelasi Heatmap dari Dataframe baru')
sns.heatmap(true_df.corr(), annot=True, cmap='cool')

Berikutnya adalah membagi data tersebut menjadi X dan y. X akan dipilih kolom distance, surge_multiplier, cab_type, windgust, windbearing, visibility, dan moonphase,  sedangkan y adalah price.

In [ ]:
X = true_df.drop(columns=['price'])
y = true_df['price']
display(X.head(3))
display(y.head(3))

In [ ]:
def histo(newdf, judul, labelx, labely='Banyaknya Index'):
    newdf.hist(grid=False, figsize=(15,7), facecolor='crimson')
    plt.title(judul)
    plt.xlabel(labelx)
    plt.ylabel(labely)
    plt.show()

Menampilkan grafik histogram pada distance

In [ ]:
histo(df['distance'], judul="Grafik Persebaran Distance" , labelx='Jarak')

Menampilkan grafik histogram pada Surge Multiplier

In [ ]:
histo(df['surge_multiplier'], judul="Grafik Persebaran Surge Multiplier" , labelx='Surge Multiplier')

Menampilkan grafik histogram pada Cab Type

In [ ]:
histo(df['cab_type'], judul="Grafik Persebaran Cab Type" , labelx='Cab Type')

Menampilkan grafik histogram pada Price

In [ ]:
histo(df['price'], judul="Grafik Persebaran Price" , labelx='Price')

In [ ]:
histo(df['visibility'], judul="Grafik Persebaran visibility" , labelx='visibility')

In [ ]:
histo(df['moonPhase'], judul="Grafik Persebaran moonPhase" , labelx='moonPhase')

In [ ]:
histo(df['windGust'], judul="Grafik Persebaran windGust" , labelx='windGust')

In [ ]:
histo(df['windBearing'], judul="Grafik Persebaran windBearing" , labelx='windBearing')

Tahap berikutnya adalah melakukan split data. Menurut dugaan saya, data akan bagus jika X distandarisasi. Akan tetapi, saya cukup ragu sehingga langkah berikutnya adalah melakukan split data dengan 2 X yaitu X yang belum distandarisasi dan X yang telah distandarisasi.

## X tanpa Standarisasi

Berikutnya X dan y akan dibagi menjadi X_train, X_test, y_train, dan y_test. Untuk random state yang dipakai adalah 69 dan train size yang digunakan adalah 69%

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.69 ,random_state=69)

print('Data X Train')
display(X_train.head(3))

print('Data X Test')
display(X_test.head(3))

print('Data y Train')
display(y_train.head(3))

print('Data y Test')
display(y_test.head(3))

## X dengan Standarisasi

Data X akan dilakukan normalisasi supaya memperbagus hasil regresi linear

In [ ]:
scaler = StandardScaler()
X_norm = scaler.fit_transform(X)
X_norm = pd.DataFrame(data=X_norm, columns=['distance', 'surge_multiplier', 'cab_type','visibility', 'windBearing', 'moonPhase', 'windGust'])
X_norm

In [ ]:
X_train_norm, X_test_norm, y_train, y_test = train_test_split(X_norm, y, train_size=0.69 ,random_state=69)

print('Data X Train')
display(X_train_norm.head(3))

print('Data X Test')
display(X_test_norm.head(3))

print('Data y Train')
display(y_train.head(3))

print('Data y Test')
display(y_test.head(3))

# Pendefinisian Model

Alasan menggunakan model regresi linear karena tujuan akhir dari projek ini. 
Tujuan projek ini yaitu memprediksi berapa harga dari lyft atau uber jika kita memasukkan jarak, surge multplier, cab, windgust, windbearing, moonphase, dan visibility.
Regresi linear merupakan algoritma machine learning yang sangat cocok untuk data yang tipikalnya continu.

# Pelatihan Model

Model pertama dibuat dengan X tanpa standarisasi

In [ ]:
model = LinearRegression().fit(X_train, y_train)

In [ ]:
y_predict = model.predict(X_test)

Model kedua X dibuat dengan standarisasi

In [ ]:
model2 = LinearRegression().fit(X_train_norm, y_train)

In [ ]:
y_predict2 = model.predict(X_test_norm)

Kita akan mencoba menggunakannya dengan data baru yang telah kami siapkan

# Evaluasi Model

Lakukan perbandingan data agar bisa mengetahui hasil prediksi yang lebih bagus

In [ ]:
y_dataframe = pd.DataFrame({
    'y_test':y_test, 
    'y_predict':y_predict, 
    'y_predict2':y_predict2,
    'yt - yp1':abs(y_test-y_predict),
    'yt - yp2':abs(y_test-y_predict2),
    'yp1 > yp2': abs(y_test-y_predict) > abs(y_test-y_predict2),
})
display(y_dataframe.head(3))
print('Ada berapa selisih ytest dengan y_predict lebih besar daripada ytest dengan y_predict2 ?')
yp1_less = y_dataframe[y_dataframe['yp1 > yp2'] == False]['yp1 > yp2'].count()
display(yp1_less)
print('Ada berapa selisih ytest dengan y_predict lebih kecil daripada ytest dengan y_predict2 ?')
yp1_more = y_dataframe[y_dataframe['yp1 > yp2'] == True]['yp1 > yp2'].count()

display(yp1_more)

Tampilan perbandingan dalam bentuk grafik

In [ ]:
data_01 = np.array([yp1_less, yp1_more])
predict_less_more = pd.DataFrame(data=data_01, columns=['False'])
predict_less_more['True'] = yp1_more
predict_less_more.drop(index=1, inplace=True)
predict_less_more

predict_less_more.plot(kind='bar', align='edge', figsize=(14,7))
plt.title('Apakah yp1 > yp2')
plt.ylabel('Banyaknya data')
plt.xlabel('Y Predict')
plt.show()

Berdasarkan data tersebut, dapat diketahui kalau tanpa adanya standarisasi, data tersebut lebih bagus daripada menggunakan standarisasi. Hal ini terjadi karena pada versi sebelum standarisasi, data X-nya masih murni sehingga koefisien dari linear regresinya akan lebih akurat untuk memprediksi hasil y yaitu price.

In [ ]:
# ypredpred = model.predict(df)
# ypredpred

Melakukan ekspor model

In [ ]:
import pickle
pickle.dump(model, open('linear_regression.pkl','wb'))

# Penarikan Kesimpulan

Beberapa informasi yang bisa didapatkan dari percobaan ini yaitu :
1. Banyaknya data adalah 693071 dan banyaknya kolom adalah 57. Akan tetapi, banyaknya data yang terpakai adalah 637976 dan banyaknya kolom yang terpakai adalah distance,surge_multiplier, price, cab_type, windgust, windbearing, visibility, dan moonphase
2. Data yang digunakan tidak terlalu bersih sehingga perlu dilakukan penanganan terhadap data yang hilang.
3. Nilai kolom cab_type diubah, tetapi tidak digunakan
4. Banyaknya kolom pada suatu data tidak harus dijadikan sebagai bagian dari linear regresi. Sebagai contoh, banyaknya berbagai kolom tentang temperature dicantumkan di dataset, tetapi tidak ada yang digunakan karena nilai korelasinya hampir mendekati 0 pada price. Pada pengukuran ini, kolom yang dipakai adalah kolom yang nilai korelasinya 0.001 pada price
5. Melakukan pemilihan kolom yang dipakai sangat berguna untuk meringankan kinerja komputasi data.
6. Pemilihan kolom tidak bisa dilakukan secara sembarang, harus ada alasan atau bukti yang melandasi. Dalam percobaan ini, bukti yang melandasi adalah type dari data dan nilai korelasi kepada price.
7. Pada percobaan kali ini, menggunakan dua jenis X yaitu X tanpa standarisasi dan X dengan standarisasi.
8. Kedua X itu membuat y predict dan hasilnya X tanpa standarisasi yang hampir mendekati y test.

Kesimpulan dalam percobaan ini yaitu :
1. Tidak selamanya data X hasil standarisasi akan lebih baik daripada X tanpa standarisasi dalam kasus linear regresi. Oleh karena itu, harus ada perbandingan terlebih dahulu.
2. Kolom dari dataset yang dipakai untuk percobaan ini adalah distance,surge_multiplier, price, cab_type, windgust, windbearing, visibility, dan moonphase. Alasannya memakai data tersebut adalah nilai korelasinya di atas 0.001 terhadap price.